kv/data_migration.go

```diff
@@ -0,0 +1,117 @@
+package kv
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/influxdata/influxdb"
+)
+
+var (
+	migrationK = []byte("result")
+	migrationV = []byte{0x1}
+)
+
+var _ influxdb.DataMigrationService = (*Service)(nil)
+
+// CheckAndMigrate will check all applicable migration and perform the migration.
+func (s *Service) CheckAndMigrate(ctx context.Context) (err error) {
+	if !s.IsBucketMigrated(ctx) {
+		if err = s.ConvertBucketToNew(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IsBucketMigrated will determine if data already migrated.
+func (s *Service) IsBucketMigrated(ctx context.Context) bool {
+	if err := s.kv.View(ctx, func(tx Tx) error {
+		b, err := tx.Bucket(influxdb.BucketIsMigratedIndex)
+		if err != nil {
+			return err
+		}
+		v, err := b.Get(migrationK)
+		if err != nil {
+			return err
+		}
+		if string(migrationV) != string(v) {
+			return &influxdb.Error{
+				Msg: "unexpected error bucket conversion error",
+			}
+		}
+		return nil
+	}); err != nil {
+		return false
+	}
+	return true
+}
+
+// ConvertBucketToNew to do a scan to the storage and convert every thing related.
+func (s *Service) ConvertBucketToNew(ctx context.Context) error {
```

Comment:
I'm concerned about this pattern for cloud. I imagine we're going to need a more sophisticated migration strategy as we will likely have to have instances that are using both the new schema and the old schema concurrently. Additionally, this is likely contain too many operations to do in a single etcd transaction.